testing/src/main/java/io/grpc/internal/testing/AbstractTransportTest.java

```diff
@@ -0,0 +1,907 @@
+    serverStreamCreation.stream.close(Status.OK, new Metadata());
+    verify(mockClientStreamListener, timeout(1000))
+        .closed(statusCaptor.capture(), any(Metadata.class));
+    assertCodeEquals(Status.OK, statusCaptor.getValue());
+  }
+
+  @Test
+  public void basicStream() throws Exception {
+    server.start(serverListener);
+    client.start(mockClientTransportListener);
+    MockServerTransportListener serverTransportListener
+        = serverListener.takeListenerOrFail(1000, TimeUnit.MILLISECONDS);
+    serverTransport = serverTransportListener.transport;
+
+    Metadata clientHeaders = new Metadata();
+    clientHeaders.put(asciiKey, "client");
+    clientHeaders.put(asciiKey, "dupvalue");
+    clientHeaders.put(asciiKey, "dupvalue");
+    clientHeaders.put(binaryKey, "äbinaryclient");
+    ClientStream clientStream = client.newStream(methodDescriptor, clientHeaders);
+    clientStream.start(mockClientStreamListener);
+    StreamCreation serverStreamCreation
+        = serverTransportListener.takeStreamOrFail(1000, TimeUnit.MILLISECONDS);
+    assertEquals(methodDescriptor.getFullMethodName(), serverStreamCreation.method);
+    assertEquals(Lists.newArrayList(clientHeaders.getAll(asciiKey)),
+        Lists.newArrayList(serverStreamCreation.headers.getAll(asciiKey)));
+    assertEquals(Lists.newArrayList(clientHeaders.getAll(binaryKey)),
+        Lists.newArrayList(serverStreamCreation.headers.getAll(binaryKey)));
+    ServerStream serverStream = serverStreamCreation.stream;
+    ServerStreamListener mockServerStreamListener = serverStreamCreation.listener;
+
+    serverStream.request(1);
+    verify(mockClientStreamListener, timeout(1000)).onReady();
+    assertTrue(clientStream.isReady());
+    clientStream.writeMessage(methodDescriptor.streamRequest("Hello!"));
+    clientStream.flush();
+    verify(mockServerStreamListener, timeout(1000)).messageRead(inputStreamCaptor.capture());
+    assertEquals("Hello!", methodDescriptor.parseRequest(inputStreamCaptor.getValue()));
+    inputStreamCaptor.getValue().close();
+
+    clientStream.halfClose();
+    verify(mockServerStreamListener, timeout(1000)).halfClosed();
+
+    Metadata serverHeaders = new Metadata();
+    serverHeaders.put(asciiKey, "server");
+    serverHeaders.put(asciiKey, "dupvalue");
+    serverHeaders.put(asciiKey, "dupvalue");
+    serverHeaders.put(binaryKey, "äbinaryserver");
+    serverStream.writeHeaders(serverHeaders);
+    verify(mockClientStreamListener, timeout(1000)).headersRead(metadataCaptor.capture());
+    assertEquals(Lists.newArrayList(serverHeaders.getAll(asciiKey)),
+        Lists.newArrayList(metadataCaptor.getValue().getAll(asciiKey)));
+    assertEquals(Lists.newArrayList(serverHeaders.getAll(binaryKey)),
+        Lists.newArrayList(metadataCaptor.getValue().getAll(binaryKey)));
+
+    clientStream.request(1);
+    verify(mockServerStreamListener, timeout(1000)).onReady();
```

Comment:
I find that to be an antipattern. For me, "helper methods" in tests tend to make the tests hard to read. I've also seen a lot of broken tests because they reused a helper method that was inappropriate. There are times it is necessary because of how much boilerplate is necessary or because the method name is more descriptive than some obscure code, but in general I would strongly prefer the duplication.